packages/grafana-runtime/src/services/dashboard-srv/DashboardSrvSingleton.ts

```diff
@@ -0,0 +1,98 @@
+import { PluginsAPIPanelModelWrapper } from './PanelWrapper';
+import {
+  CoreDashboardModel,
+  CoreDashboardSrv,
+  CorePanelModel,
+  PluginsAPIDashboardModel,
+  PluginsAPIDashboardSrv,
+  PluginsAPIPanelModel,
+} from './types';
+
+export class PluginsAPIDashboardModelWrapper implements PluginsAPIDashboardModel {
+  #dashboard: CoreDashboardModel | undefined;
+
+  // store panel references in a weakmap to avoid memory leaks
+  #panelsMap = new WeakMap<PluginsAPIPanelModel, CorePanelModel>();
+  setCurrentDashboard(dashboard: CoreDashboardModel) {
+    this.#dashboard = dashboard;
+  }
+
+  get uid() {
+    return this.#dashboard?.uid ?? '';
+  }
+
+  get title() {
+    return this.#dashboard?.title ?? '';
+  }
+
+  get panels() {
+    return this.getPanels();
+  }
+
+  getPanels() {
+    if (!this.#dashboard) {
+      return [];
+    }
+
+    // reset the weakmap
+    this.#panelsMap = new WeakMap();
+
+    //return a wrapper for each panel
+    return this.#dashboard.panels.map((panel) => {
+      const panelWrapper = new PluginsAPIPanelModelWrapper(panel);
+      // store the original panel in a symbol that can't be accessed directly
+      this.#panelsMap.set(panelWrapper, panel);
+      return panelWrapper;
+    });
+  }
+
+  updatePanels(panels: PluginsAPIPanelModel[]) {
+    if (!this.#dashboard) {
+      return;
+    }
+    const panelsToSet: CorePanelModel[] = panels.map((panel) => {
+      // if it is a panel wrapper, unwrap it
+      const originalPanel = this.#panelsMap.get(panel);
+      if (originalPanel) {
+        return originalPanel;
+      }
+      // a custom panel object from user-input
+      if (isCorePanelModel(panel)) {
+        return panel;
+      }
```

Comment:
Is this only possible if they use the old angular dashboardSrv to get the real panel model instances?

Reply:
@torkelo no. They can also fetch the dashboard and panels json directly from the http api and pass the full objects you get from there here.